### PR TITLE
feat(treesitter): add next, prev sibling method

### PR DESCRIPTION
**Pull request by Dkendal**
Add tsnode methods to change to the next, previous, named or unnamed
nodes.

Feedback much appreciated, most of this is copying and pasting in the C code what looked right :). I'm also not too sure what the process is for authoring vim documentation. 

Relates to https://github.com/neovim/neovim/issues/15408

### CODE REVIEW

**Comment by vigoux** [comment]
Hey!
Welcome in the contributor community, let me do a little review to see if we're good to go!

File: src/nvim/lua/treesitter.c [hunk 1]
```diff
@@ -992,6 +996,50 @@ static int node_parent(lua_State *L)
...
+  }
+  TSNode sibling = ts_node_next_sibling(node);
+  push_node(L, sibling, 1);
+  return 1;
+}
+
+static int node_prev_sibling(lua_State *L)
+{
+  TSNode node;
+  if (!node_check(L, 1, &node)) {
```
Comment (vigoux) [inline src/nvim/lua/treesitter.c:1013]: I think you need to do `lua_upvalue_index` here:
```suggestion
  if (!node_check(L, lua_upvalue_index(1), &node)) {
```

**Review by vigoux — COMMENTED** [review]
Hey just a little remark: could you rename the test file to `node_spec.lua`?

File: src/nvim/lua/treesitter.c (hunk 1 above)
Comment (bfredl) [reply src/nvim/lua/treesitter.c:1013]: but why

Comment (vigoux) [reply src/nvim/lua/treesitter.c:1013]: I don't really know, it feels more "robust" for some weird reason in my head.

Comment (bfredl) [reply src/nvim/lua/treesitter.c:1013]: but it is not an upvalue.

Comment (vigoux) [reply src/nvim/lua/treesitter.c:1013]: Indeed, it wasn't clear for me what an upvalue is.

**Comment by Dkendal** [comment]
> Hey just a little remark: could you rename the test file to `node_spec.lua`?

Yep no problem! I wanted to make this a unit test as well, but didn't have the time to figure out how to make that work, would that be something worth pursuing?

File: src/nvim/lua/treesitter.c (hunk 1 above)
Comment (Dkendal) [reply src/nvim/lua/treesitter.c:1013]: I'm new to writing lua C bindings so I'll double check the docs on this.

Comment (bfredl) [reply src/nvim/lua/treesitter.c:1013]: upvalues is for c closures. This is a method, so `self` is passed on stack position 1 (just as first arg).

**Comment by bfredl** [comment]
Lua api:s are tested using functionaltests (note: "unit test" in our repo means unit test of internal c functions only, "functionaltest" is everything else which is not this)

File: test/functional/treesitter/node_spec.lua [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+local clear = helpers.clear
+local eq = helpers.eq
+local exec_lua = helpers.exec_lua
+local insert = helpers.insert
+local pending_c_parser = helpers.pending_c_parser
+
+before_each(clear)
+
+local function eval(lua_expr)
```
Comment (bfredl) [inline test/functional/treesitter/node_spec.lua:11]: I would call this `lua_eval` to not confuse it with `helpers.eval`

**Review by bfredl — APPROVED** [review]

**Review by vigoux — APPROVED** [review]

**Pull request merged**